// generator-jhipster/dist/generators/base/command.mjs
const command = {
    options: {
        useVersionPlaceholders: {
            description: 'replace mutable versions with placeholders',
            type: Boolean,
            env: 'VERSION_PLACEHOLDERS',
            scope: 'generator',
            hide: true,
        },
        skipChecks: {
            description: 'Check the status of the required tools',
            type: Boolean,
            scope: 'generator',
        },
        experimental: {
            description: 'Enable experimental features. Please note that these features may be unstable and may undergo breaking changes at any time',
            type: Boolean,
            scope: 'generator',
        },
        debugEnabled: {
            name: 'debug',
            description: 'Enable debugger',
            alias: 'd',
            type: Boolean,
            scope: 'generator',
        },
        skipJhipsterDependencies: {
            description: "Don't write jhipster dependencies to package.json.",
            type: Boolean,
            scope: 'storage',
        },
    },
};
export default command;
