// generator-jhipster/dist/generators/java/support/files.mjs
/**
 * Copyright 2013-2023 the original author or authors from the JHipster project.
 *
 * This file is part of the JHipster project, see https://www.jhipster.tech/
 * for more information.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
export const replaceFilePathVariables = (data, filePath) => filePath?.replace(/_package_/, data.packageName)?.replace(/_\w*/, '');
export const replaceEntityFilePathVariables = (data, filePath) => {
    filePath = filePath
        ?.replace(/_package_/, data.entityJavaPackageFolder)
        ?.replace(/_PersistClass_/, data.persistClass)
        ?.replace(/_EntityClass_/, data.entityClass)
        ?.replace(/_DtoClass_/, data.dtoClass);
    return filePath?.includes('.jhi.') ? filePath : filePath?.replace(/_\w*/, '');
};
/**
 * Move the template to `javaPackageSrcDir` (defaults to`src/main/java/${packageFolder}/${filePath}`).
 * Removes trailing specifiers.
 */
export const moveToJavaPackageSrcDir = (data, filePath) => `${data.javaPackageSrcDir}${replaceFilePathVariables(data, filePath) ?? ''}`;
/**
 * Move the template to `javaPackageTestDir` (defaults to`src/main/java/${packageFolder}/${filePath}`).
 * Removes trailing specifiers.
 */
export const moveToJavaPackageTestDir = (data, filePath) => `${data.javaPackageTestDir}${replaceFilePathVariables(data, filePath) ?? ''}`;
export const moveToJavaEntityPackageSrcDir = (data, filePath) => `${data.srcMainJava}${data.entityAbsoluteFolder}${replaceEntityFilePathVariables(data, filePath) ?? ''}`;
export const moveToJavaEntityPackageTestDir = (data, filePath) => `${data.srcTestJava}${data.entityAbsoluteFolder}${replaceEntityFilePathVariables(data, filePath) ?? ''}`;
export const moveToSrcMainResourcesDir = (data, filePath) => `${data.srcMainResources}${replaceEntityFilePathVariables(data, filePath) ?? ''}`;
export const moveToSrcTestResourcesDir = (data, filePath) => `${data.srcTestResources}${replaceEntityFilePathVariables(data, filePath) ?? ''}`;
