// generator-jhipster/dist/generators/spring-cache/files.mjs
import { moveToJavaPackageSrcDir, moveToJavaPackageTestDir } from '../java/support/index.mjs';
import { SERVER_MAIN_SRC_DIR, SERVER_TEST_SRC_DIR } from '../generator-constants.mjs';
const files = {
    cacheFiles: [
        {
            condition: data => data.buildToolGradle,
            templates: ['gradle/cache.gradle'],
        },
        {
            path: `${SERVER_MAIN_SRC_DIR}package/`,
            renameTo: moveToJavaPackageSrcDir,
            templates: ['config/CacheConfiguration.java'],
        },
        {
            condition: data => data.cacheProviderInfinispan,
            path: `${SERVER_MAIN_SRC_DIR}package/`,
            renameTo: moveToJavaPackageSrcDir,
            templates: ['config/CacheFactoryConfiguration.java'],
        },
        {
            condition: data => data.cacheProviderRedis,
            path: `${SERVER_TEST_SRC_DIR}package/`,
            renameTo: moveToJavaPackageTestDir,
            templates: [
                'config/EmbeddedRedis.java',
                'config/RedisTestContainer.java',
                'config/RedisTestContainersSpringContextCustomizerFactory.java',
            ],
        },
    ],
};
export default async function writeTask({ application }) {
    await this.writeFiles({
        sections: files,
        context: application,
    });
}
